orders 基建

Neal-Y opened the pull request:
### CREATE
curl -X POST \
  http://localhost:8080/api/orders \
  -H "Content-Type: application/json" \
  -d '{
        "user_id": 1,
        "note": "This is a test order",
        "order_details": [{
            "product_id": 1,
            "quantity": 2
        }]
}'

### GET_FROM_ID
curl -X GET http://localhost:8080/api/orders/1

### UPDATE
curl -X PATCH \
  http://localhost:8080/api/admin/orders/1 \
  -H "Content-Type: application/json" \
  -d '{
        "note": "Updated note"
    }'

### DELETE
curl -X DELETE http://localhost:8080/api/orders/1

### GET_ALL_ORDERS
curl -X GET http://localhost:8080/api/orders

Neal-Y committed:
feat(entity) order & order_detail entity layer

Neal-Y committed:
feat(controller) order & order_detail controller layer

Neal-Y committed:
feat(controller) order & order_detail router

Neal-Y committed:
rebase test

Neal-Y committed:
fix(migration)刪除order，讓order_detail可以自動刪掉

Neal-Y committed:
feat(controller) order crud draft

Neal-Y committed:
fix(controller)
1.判斷商品有效時間成立order
2.判斷商品庫存量成立order
3.計算數量Ｘproduct單價 ＝ order_detail total_price
4.order_detail price等於product price
5.刪除訂單，product stock加回來
6.限制下單時，order_detail 的quantity不得為0

Neal-Y committed:
rebase test2

Neal-Y committed:
feat(entity)將對資料庫的CRUD移到repository layer

Neal-Y committed:
feat(entity)新增request body的dto

Neal-Y committed:
feat(repository)新增repository layer

Neal-Y committed:
feat(service)新增service layer

Neal-Y committed:
feat(handler)簡化handler流程，將業務邏輯移到service layer

Neal-Y committed:
feat(handler)新增orderService初始化

Neal-Y committed:
feat(handler)把原本的order.go的crud給他們各自一個file

Neal-Y commented:
[025a870](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/025a8703b9361554e427b26378c33d5c5cdbcfac)

Neal-Y commented on service/order.go:
```diff
@@ -0,0 +1,179 @@
+		Status:       "pending",
+		CreatedAt:    time.Now(),
+		UpdatedAt:    time.Now(),
+		OrderDetails: orderRequest.OrderDetails,
```
像是這樣的，我是不是也可以比照之前user 使用builder pattern的方式去做？

jonyig replied:
用 `builder` or `orderRequest.exportOrderModel()`
都可以

Neal-Y replied:
[2570235](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/2570235f825c928319e93cf91d8321753f3fad8f)

Neal-Y committed:
feat(builder)新增order builder

Neal-Y committed:
刪掉order builder

Neal-Y committed:
feat(entity)order dto 增加status

Neal-Y committed:
fix(service)修復update api中以下幾點：
1.保留原有的 total_price 值：如果請求中提供了訂單詳細信息，則重新計算 total_price；否則保留原有的 total_price。
2.條件更新 note 字段：僅在請求中提供了 note 字段時才更新 note。
3.條件更新 status 字段：僅在請求中提供了 status 字段時才更新 status。
4.保留原有的 orderDetails：如果請求中沒有提供訂單詳細信息，則保留原有的 orderDetails。

jonyig commented on database/migrations/20240527063608_create_order_detail_table.up.sql:
```diff
@@ -6,6 +6,6 @@ CREATE TABLE order_details (
     created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
     updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
-    FOREIGN KEY (order_id) REFERENCES orders(id),
+    FOREIGN KEY (order_id) REFERENCES orders(id) ON DELETE CASCADE ON UPDATE CASCADE,
```
database 相關改動，是新增一個 migration ，不是改舊的

Neal-Y replied:
我想說我目前還沒有用戶資料，所以我才直接這樣去洗掉資料庫做更改，而不是用alter

Neal-Y replied:
123

jonyig replied:
OK, 以後盡量不要

jonyig commented on model/database/orders.go:
```diff
@@ -0,0 +1,37 @@
+	OrderDetails []OrderDetail `json:"order_details,omitempty" gorm:"foreignKey:OrderID"`
+}
+
+type OrderDetail struct {
```
拆成兩個 file 吧

Neal-Y replied:
[21dd979](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/21dd97933e4d1fcf7efd20af691eb11d40d0156d)

jonyig replied:
123

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,188 @@
+func (s *orderService) CreateOrder(orderRequest *datatransfer.OrderRequest) (*database.Order, error) {
+	totalPrice := 0.0
+
+	for i, detail := range orderRequest.OrderDetails {
```
這邊可以包裝成 function ，類似於 order validation

Neal-Y replied:
[27970fa](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/27970fa65df113117e30e2760630e1fc1c618b52)

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,188 @@
+		return nil, err
+	}
+
+	for _, detail := range order.OrderDetails {
```
這邊個 sql execute 會做 n * 2 次

假設有  4 個 detail 會執行 8 次

4 find + 4 update

可以嘗試縮小到 2 次

Neal-Y replied:
[27970fa](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/27970fa65df113117e30e2760630e1fc1c618b52)

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,188 @@
+	return s.orderRepo.FindByID(id)
+}
+
+func (s *orderService) UpdateOrder(id int, orderRequest *datatransfer.OrderRequest) (*database.Order, error) {
```
這邊我沒往下看

因為要先定義出允許修改哪些值

我覺得不需開放修改 order detail

Neal-Y replied:
[27970fa](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/27970fa65df113117e30e2760630e1fc1c618b52)

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,188 @@
+		}
+	}
+
+	err = s.orderRepo.Delete(order)
```
可以使用 soft delete

Neal-Y replied:
[27970fa](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/27970fa65df113117e30e2760630e1fc1c618b52)

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,188 @@
+	if len(orderRequest.OrderDetails) > 0 {
+		totalPrice = 0.0
+		for _, detail := range orderRequest.OrderDetails {
+			product, err := s.productRepo.FindByID(detail.ProductID)
```
這邊可以跟 95 行一起使用 join 拉出來，就不用打很多次 DB

Neal-Y replied:
[27970fa](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/27970fa65df113117e30e2760630e1fc1c618b52)

這部分我們更改掉了，就是決定updateorder API只開放admin修改status以及note

Neal-Y committed:
feat(entity)更改update api所用到的dto，讓更新變得簡單明瞭

Neal-Y committed:
feat(entity)將order兩個table分開，並且新增單獨放order_detail table的file

Neal-Y committed:
feat(entity)更改dto的名稱

Neal-Y committed:
feat(repository)增加order的軟刪除，也就是將訂單狀態改成"cancelled"

Neal-Y committed:
feat(repository)新增批量上傳的方式，使得更新庫存時，可以在查詢和更新的次數降到兩次

Neal-Y committed:
feat(service)更新如下：
1.新增批量上傳的方式，使得更新庫存時，可以在查詢和更新的次數降到兩次
2.新增validateOrderRequest函數來處理，createOrder時的驗證
3.限制updateOrder只有admin可以使用，並且只能update 狀態或是註記
4.商品售罄

Neal-Y committed:
feat(handler)更改dto 名稱

Neal-Y committed:
feat(handler)更改dto 名稱和使用不同update函數

Neal-Y committed:
feat(handler)用prefix切分區別，更新只有admin可以使用

Neal-Y committed:
由於dto都切開成各自的file，所以更改相依賴file裡的”路徑“和“名稱”

jonyig commented on service/order.go:
```diff
@@ -28,25 +29,26 @@ func NewOrderService(orderRepo repository.OrderRepository, productRepo repositor
+			return 0, nil, errors.New("Quantity must be greater than zero")
 		}
 
 		product, err := s.productRepo.FindByID(detail.ProductID)
```
這邊還是存在 N+1 issue 

https://segmentfault.com/a/1190000039421843

Neal-Y replied:
[282a544](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/282a544562cb60cd5dd7fa3a758fd516579d4022)
這是repository/product.go
[72cbb7b](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/72cbb7b96ca6acbd59d244ba64a70719cbc4f8d9)
這是service/order.go

service/order.go引用repository/product.go中的FindByIDs()來解決n+1問題，並且用repository/product.go中的BatchUpdate()批量更新回product table

jonyig commented on service/order.go:
```diff
@@ -0,0 +1,161 @@
+	UpdateOrderStatusAndNote(id int, orderRequest *order.StatusRequest) (*database.Order, error)
+	DeleteOrder(id int) error
+	ListAllOrders() ([]database.Order, error)
+	validateOrderRequest(orderRequest *order.Request) (float64, map[int]*database.Product, error)
```
validateOrderRequest 設定成 private ，應該不需要放到 interface 上面吧

Neal-Y replied:
[72cbb7b](https://github.com/Neal-Y/devmentor-BE103-golang/pull/7/commits/72cbb7b96ca6acbd59d244ba64a70719cbc4f8d9)

Neal-Y committed:
feat(product repository)新增FindByIDs，我將需要更新的product id全部塞到一個slice中，並且傳到這個函數去進行where in的查詢

Neal-Y committed:
feat(service)更改如下：
1.validateOrderRequest函數由於是CreateOrder內部使用的，所以移出OrderService interface
2.解決重複查詢的product的n+1問題，一次查出來塞到products變數中
3.CreateOrder塞入“事務”，避免更新失敗，有髒資料
4.新增感覺只是比較美觀的builder pattern到CreateOrder中

Neal-Y committed:
feat(builder)order service的builder

Neal-Y committed:
feat(service)ㄎㄧㄤ 掉忘記塞rollback，事務個寂寞
